File: frontend/mobile/lib/models/responses/ticket_validation_response.dart
/// The answer to one scan (`POST /api/tickets/validate`).
///
/// Arrives as a **200** even when [isValid] is false: "already used", "wrong
/// event", "expired" are displayable answers about a ticket, not transport
/// failures, and the scanner has to paint a red card with [message] rather
/// than a generic error toast. Genuine failures — 403 (not an organizer) and
/// 409 (another device is validating this same ticket right now) — still come
/// through as an `ApiException`.
///
/// [message] is authored server-side in Bosnian and rendered verbatim, so the
/// reason a person is turned away at the door can never drift between the API
/// and this screen.
class TicketValidationResponse {
  final bool isValid;
  final String code;
  final String message;
  final String? ticketId;
  final String? sectorName;
  final String? ticketTypeName;
  final String? holderEmail;
  final DateTime? validatedAt;

  const TicketValidationResponse({
    required this.isValid,
    required this.code,
    required this.message,
    this.ticketId,
    this.sectorName,
    this.ticketTypeName,
    this.holderEmail,
    this.validatedAt,
  });

  factory TicketValidationResponse.fromJson(Map<String, dynamic> json) {
    return TicketValidationResponse(
      isValid: json['isValid'] as bool? ?? false,
      code: json['code'] as String? ?? '',
      message: json['message'] as String? ?? '',
      ticketId: json['ticketId'] as String?,
      sectorName: json['sectorName'] as String?,
      ticketTypeName: json['ticketTypeName'] as String?,
      holderEmail: json['holderEmail'] as String?,
      validatedAt: json['validatedAt'] != null
          ? DateTime.tryParse(json['validatedAt'] as String)
          : null,
    );
  }
}

File: backend/services/ticketing/eTicketing.Ticketing.Business/Tickets/Dtos/TicketValidationResponse.cs
namespace eTicketing.Ticketing.Business.Tickets;

/// <summary>
/// Deliberately returned as a SUCCESSFUL Result even when <paramref name="IsValid"/> is false. A
/// ticket that's already been used, belongs to another event, or expired is an expected, displayable
/// answer — the scanner has to paint a red card with the reason, which it can't do if the transport
/// swallowed it as a 404. Genuine failures (caller isn't an organizer → 403, lost the Redis lock →
/// 409) stay Result.Failure.
/// </summary>
/// <param name="Direction">What this particular scan did. <see cref="GatePassageDirection.Entry"/>
/// for every admitted one-shot ticket and for a subscription holder coming in;
/// <see cref="GatePassageDirection.Exit"/> for a subscription holder going out — a valid scan that
/// must NOT read as "admitted" on the scanner's green card.</param>
/// <param name="IsInside">RecurringReservation only: whether the holder is inside after this scan.
/// Always false for the one-shot modes, which have no such state.</param>
public sealed record TicketValidationResponse(
    bool IsValid,
    string Code,
    string Message,
    Guid? TicketId,
    string? SectorName,
    string? TicketTypeName,
    string? HolderEmail,
    DateTime? ValidatedAt,
    GatePassageDirection Direction = GatePassageDirection.None,
    bool IsInside = false);
